### PR TITLE
hcm: move ownership of headers/trailers into FM

### PR DESCRIPTION
**Pull request by snowp**
This moves the ownership over the headers/trailers from the ActiveStream to the FilterManager, providing accessors
that allow the AS to access the headers. 

Risk Level: Medium
Testing: Existing tests
Docs Changes: n/a
Release Notes: n/a
Part of #10454

### CODE REVIEW

**Comment by mattklein123** [comment]
CI failure looks legit.

/wait

File: source/common/http/conn_manager_impl.cc [hunk 1]
```diff
@@ -1594,23 +1607,26 @@ void ConnectionManagerImpl::ActiveStream::sendLocalReply(
...
+                                                               body, content_type);
             },
             [&](ResponseHeaderMapPtr&& response_headers, bool end_stream) -> void {
               if (modify_headers != nullptr) {
                 modify_headers(*response_headers);
               }
-              response_headers_ = std::move(response_headers);
-              encodeHeaders(*response_headers_, end_stream);
+              // TODO(snowp): This is kinda awkward but we need to do this so that the access log
+              // sees these headers. Is there a better way?
```
Comment (mattklein123) [inline source/common/http/conn_manager_impl.cc:1619]: Shouldn't sendLocalReply just always get routed through the filter manager?

File: source/common/http/conn_manager_impl.h [hunk 2]
```diff
@@ -533,6 +538,37 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
...
+    void setRequestHeaders(RequestHeaderMapPtr&& request_headers) {
+      ASSERT(request_headers_ == nullptr);
+      request_headers_ = std::move(request_headers);
+    }
+
+    void setResponseHeaders(ResponseHeaderMapPtr&& response_headers) {
+      // Note: sometimes the headers get reset (local reply while response is buffering), so we
+      // don't assert here.
+      response_headers_ = std::move(response_headers);
+    }
```
Comment (mattklein123) [inline source/common/http/conn_manager_impl.h:550]: The existence of these functions seems odd to me. I already commented on response headers, but for request headers couldn't the filter manager store the request headers when it's decode headers method is called? Why does it have to be done out of band?

File: source/common/http/conn_manager_impl.cc (hunk 1 above)
Comment (snowp) [reply source/common/http/conn_manager_impl.cc:1619]: The comment above explains why we skip the filters here: 
```
    // In this case, at least the header and possibly the body has started
    // processing through the filter chain, but no non-informational headers
    // have been sent downstream. To ensure that filters don't get their
    // state machine screwed up, bypass the filter chain and send the local
    // reply directly to the codec.
```

File: source/common/http/conn_manager_impl.h (hunk 2 above)
Comment (snowp) [reply source/common/http/conn_manager_impl.h:550]: It gets a bit tricky with local replies: several code paths in `decodeHeaders` might send a local reply before we'd dispatched the request header to the filter manager. I think the main reason why we need the headers set before we do a local reply is that we want to make sure that they appear in access logs.

I think one option would be to have a `sendLocalReply` call that accepts a `RequestHeaderMapPtr&&` that takes ownership over the headers before sending a local reply, with ASSERTs verifying that we never call the other `sendLocalReply` overload without headers haven't being set via `decodeHeaders`?

File: source/common/http/conn_manager_impl.cc (hunk 1 above)
Comment (mattklein123) [reply source/common/http/conn_manager_impl.cc:1619]: I see, OK fair enough to leave this TODO for now.

File: source/common/http/conn_manager_impl.h (hunk 2 above)
Comment (mattklein123) [reply source/common/http/conn_manager_impl.h:550]: > I think one option would be to have a sendLocalReply call that accepts a RequestHeaderMapPtr&& that takes ownership over the headers before sending a local reply, with ASSERTs verifying that we never call the other sendLocalReply overload without headers haven't being set via decodeHeaders?

Yeah this sounds good. Maybe just put in TODOs for now and work on cleaning this up later once the code is split?

/wait

**Pull request merged**